### PR TITLE
Implement alive across call for CSE candidates

### PR DESCRIPTION
**Pull request by briansull**
This is a Zero-Diff change
Using the alive across call information will take place in a future change.
This information will influence how aggressively we try to replace a CSE candidate into a new CSE LclVar.

### CODE REVIEW

File: src/jit/compiler.h [hunk 1]
```diff
@@ -6088,8 +6088,9 @@ class Compiler
 
     static const int MIN_CSE_COST = 2;
 
-    // Keeps tracked cse indices
-    BitVecTraits* cseTraits;
+    // trait informatiobn for CSE bitsets
```
Comment (AaronRobinsonMSFT) [inline src/jit/compiler.h]: ~~informatiobn~~ -> information

File: src/jit/morph.cpp [hunk 2]
```diff
@@ -8245,6 +8245,8 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
         }
     }
 
+    compCurBB->bbFlags |= BBF_HAS_CALL;
```
Comment (AaronRobinsonMSFT) [inline src/jit/morph.cpp]: There are various places above where we return early. If indicating this here is okay, can you comment as to why the above doesn't count as a call?

**Comment by briansull** [comment]
@dotnet/jit-contrib PTAL

This change updates the dataflow used by optCSE to properly compute the alive across call information.  It is store per CSE candidate in csdLiveAcrossCall 

As this is a Zero diff change we don't use this information with this change.  Instead the function LiveAcrossCall() will always return false  as it currently does today.

In a future change we can tune the CSE logic to be less aggressive when the CSE candidate is alive across a call.

File: src/jit/optcse.cpp [hunk 3]
```diff
@@ -832,21 +847,94 @@ void Compiler::optValnumCSE_InitDataFlow()
...
+                    if (stmt == block->firstStmt())
+                    {
+                        break;
+                    }
+                    stmt = stmt->GetPrevStmt();
+                }
+
+                // Set the block->bbCseGen by or-ing in the afterCallGen set
+                //
+                BitVecOps::UnionD(cseTraits, block->bbCseGen, afterCallGen);
```
Comment (mikedn) [inline src/jit/optcse.cpp]: `afterCallGen` doesn't seem to be necessary - it collects CSE defs and then these defs get added to `bbCseGen`. So add the CSE defs directly to `bbCseGen`.

File: src/jit/optcse.cpp [hunk 4]
```diff
@@ -832,21 +847,94 @@ void Compiler::optValnumCSE_InitDataFlow()
...
             lst = lst->tslNext;
         }
     }
 
+    for (BasicBlock* block = fgFirstBB; block; block = block->bbNext)
+    {
+        // If the block contains a call we may have to update the bbCseGen
+        // as we may generate some CSEs after the last call in the block.
+        //
+        if ((block->bbFlags & BBF_HAS_CALL) != 0)
```
Comment (mikedn) [inline src/jit/optcse.cpp]: It would be better to make this `if (no call) continue`. It avoids additional indentation in the code below and makes it easy to see that the loop only looks at blocks with calls.

File: src/jit/optcse.cpp [hunk 5]
```diff
@@ -882,37 +970,68 @@ class CSE_DataFlow
...
     EXPSET_TP     m_preMergeOut;
+#ifdef DEBUG
+    Compiler* m_pCompiler;
+#endif // DEBUG
 
 public:
     CSE_DataFlow(Compiler* pCompiler) : m_pBitVecTraits(pCompiler->cseTraits), m_preMergeOut(BitVecOps::UninitVal())
     {
+#ifdef DEBUG
+        m_pCompiler = pCompiler;
```
Comment (mikedn) [inline src/jit/optcse.cpp]: Constructors should initialize data members in the initialization list.

File: src/jit/optcse.cpp [hunk 6]
```diff
@@ -832,21 +847,94 @@ void Compiler::optValnumCSE_InitDataFlow()
...
-            BasicBlock* block = lst->tslBlock;
-            BitVecOps::AddElemD(cseTraits, block->bbCseGen, genCSEnum2bit(CSEindex));
+            BasicBlock* block   = lst->tslBlock;
+            unsigned    cseBit  = genCSEnum2bit(CSEindex) * 2;
+            unsigned    cseBit2 = cseBit + 1;
+
+            // This CSE is generated in 'block', we always set the cseBit
+            // If this block does not contain a call, we also set cseBit2
+            //
+            // If we have a call in this block the below we walk the trees backwards
```
Comment (mikedn) [inline src/jit/optcse.cpp]: "the below loop walks" or something like that I presume

File: src/jit/optcse.cpp [hunk 7]
```diff
@@ -832,21 +847,94 @@ void Compiler::optValnumCSE_InitDataFlow()
...
         treeStmtLst* lst      = dsc->csdTreeList;
         noway_assert(lst);
 
         while (lst != nullptr)
         {
-            BasicBlock* block = lst->tslBlock;
-            BitVecOps::AddElemD(cseTraits, block->bbCseGen, genCSEnum2bit(CSEindex));
+            BasicBlock* block   = lst->tslBlock;
+            unsigned    cseBit  = genCSEnum2bit(CSEindex) * 2;
+            unsigned    cseBit2 = cseBit + 1;
```
Comment (mikedn) [inline src/jit/optcse.cpp]: A more descriptive name would be useful. Perhaps `cseCrossCallBit`?

File: src/jit/optcse.cpp [hunk 8]
```diff
@@ -1239,6 +1403,54 @@ void Compiler::optValnumCSE_Availablity()
...
+                        // etc..) Either isUse or isDef will be set to true in this case.
+
+                        // For a Def we will stil have a call, but it makes one CSE available so don't kill it
+                        // but still kill all the other CSEs
+                        //
+                        if (isDef)
+                        {
+
+                            EXPSET_TP call_killset = BitVecOps::UninitVal();
+                            BitVecOps::Assign(cseTraits, call_killset, call_kills_all_cses);
```
Comment (mikedn) [inline src/jit/optcse.cpp]: This will always allocate a bitset when the number of CSEs exceeds 32. You could hoist this bitset out of the loop to avoid this. But this acrobatics seems unnecessary, you could just save `cseBit2` bit from `available_cses` to a variable and restore it after doing the intersection.

File: src/jit/optcse.cpp (hunk 7 above)
Comment (briansull) [reply src/jit/optcse.cpp]: The comment above states:

```
    // two bits are allocated per CSEcandidate,
    // the first one 'cseMaskTraits' tracks normal availability
    // the second one 'cseTraits' tracks availability when calls are considered killing
```

Comment (briansull) [reply src/jit/optcse.cpp]: I am open for suggestions on what names to use here.

File: src/jit/optcse.cpp (hunk 4 above)
Comment (briansull) [reply src/jit/optcse.cpp]: OK

File: src/jit/optcse.cpp (hunk 3 above)
Comment (briansull) [reply src/jit/optcse.cpp]: OK I can add them directly to bbCseGen instead.

File: src/jit/optcse.cpp (hunk 5 above)
Comment (briansull) [reply src/jit/optcse.cpp]: OK

File: src/jit/optcse.cpp (hunk 7 above)
Comment (CarolEidt) [reply src/jit/optcse.cpp]: I think that `cseCrossCallBit` seems reasonable, as suggested by @mikedn. 
That said, I find the comment that you reference above to be confusing. You say "two bits are allocated" and then say that the second one is `cseTraits`. But AFAICT, both bits you describe are in `cseTraits`.
I haven't fully reviewed this code yet, but it seems there are actually 3 bits per CSE candidate (2 in `cseTraits` and 1 in `cseMaskTraits`). You describe the latter as reflecting "normal availability", and I *think* the former has one bit for whether the cse is defined, and another for whether it is live across a call. If that's correct it would be good to make that clear in that comment above.

Comment (briansull) [reply src/jit/optcse.cpp]: You are right and that comment is incorrect.
These traits controls how many bits get allocated for a particular Bit Vector

There are two kinds needed now when dealing with CSEs:

`cseMaskTraits` only needs one bit per CSE candidate and it is only used by `optCSE_canSwap`
for `CSE_useMask` and `CSE_defMask` for a legality check.

I will rename the other one 'cseTraits' to 'cseLivenessTraits'. It needs two bits per CSE candidate plus one extra bit so that the liveness algorithm properly visits all children on the first visit.  

Note that I discovered this bug when I investigate couple of functions that had unexpected diffs. It previously existed, but was harder to hit because we over-allocated using BitVecTraits(EXPSET_SZ, this) on line 366, so could happen only when we have the maximum number of CSEs and an appropriate initial liveness state.

Now each CSE gets two bits, where a one bit means that the CSE is available and a 0 bit means that it is not available.   So at a definition you would see both bits changing from 0 to 1 and when visiting a call the second bit would get killed and become 0, while the other availability bit stays as 1.  I am currently calling the second bit `cseBit2`

I can call it `cseCrossCallBit` but it will go from 1 to 0 when it crosses a call.

Comment (briansull) [reply src/jit/optcse.cpp]: I renamed both bits: cseBit to cseAvailBit and cseBit2 to cseAvailCrossCallBit

**Comment by briansull** [comment]
@dotnet/jit-contrib PTAL

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

**Comment by briansull** [comment]
@CarolEidt PTAL

File: src/jit/compiler.h [hunk 9]
```diff
@@ -6115,8 +6115,9 @@ class Compiler
 
     static const int MIN_CSE_COST = 2;
 
-    // Keeps tracked cse indices
-    BitVecTraits* cseTraits;
+    // trait information for CSE bitsets
+    BitVecTraits* cseMaskTraits;     // one bit per CSE candidate
+    BitVecTraits* cseLivenessTraits; // two bits per CSE candidate + 1
```
Comment (CarolEidt) [inline src/jit/compiler.h]: You should add a comment here (similar to your response to my question before) about what these are used for.

File: src/jit/optcse.cpp [hunk 10]
```diff
@@ -788,9 +798,22 @@ void Compiler::optCseUpdateCheckedBoundMap(GenTree* compare)
...
  *  Compute each blocks bbCseGen
  *  This is the bitset that represents the CSEs that are generated within the block
+ *  Also initialize bbCseIn, bbCseOut and bbCseGen sets for all blocks
  */
 void Compiler::optValnumCSE_InitDataFlow()
 {
+    // Init traits and full/empty bitvectors.  This will be used to track the
+    // individual cse indexes.
+
+    // two bits, plus one extra bit is allocated per CSEcandidate to compute CSE liveness
```
Comment (CarolEidt) [inline src/jit/optcse.cpp]: I would reword this as:
```
// We allocate two bits per CSE candidate, plus an extra bit to handle the initial unvisited case.
// The first bit tracks ...
```
And somewhere it would be good to describe how that extra bit is used to handle the unvisited case.

File: src/jit/optcse.cpp [hunk 11]
```diff
@@ -1013,23 +1187,47 @@ void Compiler::optValnumCSE_Availablity()
...
+                    //
+                    bool madeLiveAcrossCall = false;
+                    if (isUse && !desc->csdLiveAcrossCall &&
+                        !BitVecOps::IsMember(cseLivenessTraits, available_cses, cseAvailCrossCallBit))
+                    {
+                        desc->csdLiveAcrossCall = true;
+                        madeLiveAcrossCall      = true;
+                    }
+
+                    // If this a a CSE def, the the call-kill bit should also be a def (i.e not set in the
```
Comment (CarolEidt) [inline src/jit/optcse.cpp]: typo "the the" should be "then the" I think.
It is confusing to say that "the call-kill bit should also be a def". I think it would be clearer to change this comment to:
```
// If this is a CSE def (i.e. the CSE is not available here, since it is being defined), then the call-kill bit
// should also be zero since it is also not available across a call.
```

File: src/jit/optcse.cpp [hunk 12]
```diff
@@ -880,39 +972,103 @@ void Compiler::optValnumCSE_InitDataFlow()
...
+    EXPSET_TP m_preMergeOut;
+    EXPSET_TP m_callKills;
 
 public:
-    CSE_DataFlow(Compiler* pCompiler) : m_pBitVecTraits(pCompiler->cseTraits), m_preMergeOut(BitVecOps::UninitVal())
+    CSE_DataFlow(Compiler* pCompiler)
+        : m_comp(pCompiler), m_preMergeOut(BitVecOps::UninitVal()), m_callKills(BitVecOps::UninitVal())
     {
+        m_callKills = BitVecOps::MakeEmpty(m_comp->cseLivenessTraits);
+        for (unsigned inx = 0; inx < m_comp->optCSECandidateCount; inx++)
```
Comment (CarolEidt) [inline src/jit/optcse.cpp]: It seems that you should be able to factor out this code so that it isn't duplicated below.

File: src/jit/optcse.cpp [hunk 13]
```diff
@@ -880,39 +972,103 @@ void Compiler::optValnumCSE_InitDataFlow()
...
+        {
+            unsigned CseAvailBit = inx * 2;
+
+            // a one preserves availability and a zero kills the availability
+            // we generate this kind of bit pattern:  101010101010
```
Comment (CarolEidt) [inline src/jit/optcse.cpp]: If I understand correctly, these two bits really reflect 3 states:
- The CSE is available and is not killed by a call - `11`
- The CSE is available, but is killed by a call - `10`
- The CSE is not available - `00`

The `10` pattern seems counter-intuitive as an initial setting. If you want to initialize it to "full" it would seem logical to turn on both bits, and if you want to initialize it to "empty" it would seem logical to turn off both bits. What am I missing here?

Comment (briansull) [reply src/jit/optcse.cpp]: Yes, your understanding is basically correct, but I would express it this way:

`11` - The CSE is available, and is also available when considering calls as killing availability.
`10 `- The CSE is available, but is not available when considering calls as killing availability.
`00`  - The CSE is not available
`01 `- Illegal combination

This code constructs the mask `m_callKills` which is used below to clear the lower bit, - `10` of the `bbCseIn` for every CSE candidate, when the block contains a callsite:

```
            if (block->bbFlags & BBF_HAS_CALL)
            {
                BitVecOps::IntersectionD(m_comp->cseLivenessTraits, cseIn_withCallsKill, m_callKills);
            }
```

File: src/jit/optcse.cpp [hunk 14]
```diff
@@ -880,39 +972,103 @@ void Compiler::optValnumCSE_InitDataFlow()
...
+            //
+            BitVecOps::AddElemD(m_comp->cseLivenessTraits, m_callKills, CseAvailBit);
+        }
     }
 
     // At the start of the merge function of the dataflow equations, initialize premerge state (to detect changes.)
     void StartMerge(BasicBlock* block)
     {
-        BitVecOps::Assign(m_pBitVecTraits, m_preMergeOut, block->bbCseOut);
+        BitVecOps::Assign(m_comp->cseLivenessTraits, m_preMergeOut, block->bbCseOut);
```
Comment (briansull) [inline src/jit/optcse.cpp]: This initializes `m_preMergeOut` to `cseFull`, which is all bits set,  because we previous set `bbCseOut` to `cseFull` in `optValnumCSE_InitDataFlow()`

```
cseFull           = BitVecOps::MakeFull(cseLivenessTraits);
block->bbCseOut = BitVecOps::MakeCopy(cseLivenessTraits, cseFull);
```

File: src/jit/optcse.cpp [hunk 15]
```diff
@@ -880,39 +972,103 @@ void Compiler::optValnumCSE_InitDataFlow()
...
+        {
+            BitVecOps::Assign(m_comp->cseLivenessTraits, cseIn_withCallsKill, block->bbCseIn);
+            if (block->bbFlags & BBF_HAS_CALL)
+            {
+                BitVecOps::IntersectionD(m_comp->cseLivenessTraits, cseIn_withCallsKill, m_callKills);
+            }
+        }
+        BitVecOps::DataFlowD(m_comp->cseLivenessTraits, block->bbCseOut, block->bbCseGen, cseIn_withCallsKill);
+
+        bool result = !BitVecOps::Equal(m_comp->cseLivenessTraits, block->bbCseOut, m_preMergeOut);
```
Comment (briansull) [inline src/jit/optcse.cpp]: This is why we need the extra bit:

This dataflow algorithm will push all of this block successors onto the `worklist `to visit when this `EndMerge` step returns true:

In dataflow.h line 63:
```
        if (callback.EndMerge(block))
        {
            for (BasicBlock* succ : block->GetAllSuccs(m_pCompiler))
            {
                worklist.insert(worklist.end(), succ);
            }
        }
```

If the very first time that we vist this block we compute that `block->bbCseOut`, has all bits set, (i.e `cseFull`)
Then we won't visit this blocks successors at all.

Before my changes here, we always used `unsigned (-1) ` as the value for cseFull, this meant that we essentially always had one or more extra bits set in `bbCseOut`.  

I found this issue after my change because there were a couple of methods that had blocks that made every CSE available and then the algorithm didn't properly visit successor blocks.
`

File: src/jit/optcse.cpp (hunk 13 above)
Comment (briansull) [reply src/jit/optcse.cpp]: I can rename `m_callKills` to `m_callKillsMask` to make this more clear.

**Comment by briansull** [comment]
@dotnet/jit-contrib  PTAL - I have addressed all of the feedback from Carol and have tested this again, with zero diffs.  It will enable better CSE choices in a future check in.
